tests/common/plugins/sanity_check/recover.py

```diff
@@ -26,13 +27,58 @@ def reboot_dut(dut, localhost, cmd):
         assert False, "Failed to reboot the DUT"
 
     localhost.wait_for(host=dut.hostname, port=22, state="started", delay=10, timeout=300)
-    wait(30, msg="Wait 30 seconds for system to be stable.")
+    wait(120, msg="Wait 120 seconds for system to be stable.")
 
 
-def recover(dut, localhost, recover_method):
+def __recover_interfaces(dut, fanouthosts, result):
+    for port in result['down_ports']:
+        logging.info("Restoring port {}".format(port))
+        fanout, fanout_port = fanout_switch_port_lookup(fanouthosts, port)
+        if fanout and fanout_port:
+            fanout.no_shutdown(fanout_port)
+        dut.no_shutdown(port)
+    wait(30, msg="Wait 30 seconds for interface(s) to restore.")
+
+
+def __recover_services(dut, result):
+    status   = result['services_status']
+    services = [ x for x in status if not status[x] ]
+    logging.info("Service(s) down: {}".format(services))
+    return 'reboot' if 'database' in services else 'config_reload'
+
+
+def adaptive_recover(dut, localhost, fanouthosts, check_results):
+    outstanding_action = None
+    for result in check_results:
+        if result['failed']:
+            logging.info("Restoring {}".format(result))
+            if result['check_item'] == 'interfaces':
+                __recover_interfaces(dut, fanouthosts, result)
+            elif result['check_item'] == 'services':
+                action             = __recover_services(dut, result)
+                # Only allow outstanding_action be overridden when it is
+                # None. In case the outstanding_action has already been
+                # been set to 'reboot'.
+                outstanding_action = action if not outstanding_action
+            else:
+                outstanding_action = 'reboot'
+
+    if outstanding_action:
+        method = constants.RECOVER_METHODS[outstanding_action]
+        if method["reboot"]:
+            reboot_dut(dut, localhost, constants.RECOVER_METHODS[recover_method]["cmd"])
+        else:
+            dut.command(method["cmd"])
+            wait(60, msg="Wait 60 seconds for system to be stable.")
+
+
+def recover(dut, localhost, fanouthosts, check_results, recover_method):
     logger.info("Try to recover %s using method %s" % (dut.hostname, recover_method))
-    if constants.RECOVER_METHODS[recover_method]["reboot"]:
+    method = constants.RECOVER_METHODS[recover_method]
+    if method["adaptive"]:
+        adaptive_recover(dut, localhost, fanouthosts, check_results)
+    elif method["reboot"]:
         reboot_dut(dut, localhost, constants.RECOVER_METHODS[recover_method]["cmd"])
     else:
-        dut.command(constants.RECOVER_METHODS[recover_method]["cmd"])
-        wait(30, msg="Wait 30 seconds for system to be stable.")
+        dut.command(method["cmd"])
+        wait(60, msg="Wait 60 seconds for system to be stable.")
```

Comment:
Should this be refactored into a helper like the reboot_dut method now that this "run and wait" bit is being used in multiple places? I think at the least this wait period should be made a constant (e.g. CMD_RECOVERY_WAIT_PERIOD = 60).

Reply:
Made some change. What do you think?

Reply:
Looks good to me! :)